### PR TITLE
executor: group_concat aggr panic when session.group_concat_max_len is small

### LINKED ISSUE #23129: group_concat panic when @@group_concat_max_limit is small
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

    drop table s;
    create table s(a varchar(10));
    insert into s values("0123456789");
    insert into s values("12345");
    set @@group_concat_max_len=8;
    select group_concat(a order by a) from s group by a;
### 2. What did you expect to see? (Required)
    mysql> select group_concat(a order by a) from s group by a;
    +----------------------------+
    | group_concat(a order by a) |
    +----------------------------+
    | 01234567                   |
    | 12345                      |
    +----------------------------+
    2 rows in set, 1 warning (0.00 sec)
### 3. What did you see instead (Required)
    mysql>     select group_concat(a order by a) from s group by a;
    ERROR 1105 (HY000): bytes.Buffer: truncation out of range
    m
### 4. What is your TiDB version? (Required)
master(Server version: 5.7.25-TiDB-v4.0.0-beta.2-2274-gdb62e341f-dirty TiDB Server (Apache License 2.0) Community Edition, MySQL 5.7 compatible)

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #23129 

Problem Summary: 

    drop table if exists t1;
    create table t1(cid int, sname varchar(100));
    insert into t1 values(1, 'Bob'), (1, 'Alice');
    insert into t1 values(3, 'Ace');
    set @@group_concat_max_len=5;
    select group_concat(sname order by sname) from t1 group by cid;

### What is changed and how it works?

What's Changed:  when concat all strings from a group(AppendFinalResult2Chunk), we add the last separator to result, and then truncate to sesssion.group_concat_max_len. But we forgot to check if the concated string is shorter than sesssion.group_concat_max_len. So program panic.

PS: Another compatible bug found when fixing this bug, but it's not critical. Will fix in another issue.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test: Add test in aggregate_test.go::TestGroupConcatAggr()

Side effects: no

### Release note 

- group_concat aggr panic when session.group_concat_max_len is small.

### CODE REVIEW

**Comment by guo-shaoge** [comment]
Another compatible bug found(same case as this PR)

mysql 8.0:

    Server version: 8.0.23 MySQL Community Server - GPL
    
    Copyright (c) 2000, 2021, Oracle and/or its affiliates.
    
    Oracle is a registered trademark of Oracle Corporation and/or its
    affiliates. Other names may be trademarks of their respective
    owners.
    
    Type 'help;' or '\h' for help. Type '\c' to clear the current input statement.
    
    mysql> set @@group_concat_max_len=5;
    Query OK, 0 rows affected (0.00 sec)
    
    mysql> select group_concat(sname order by sname) from t1 group by cid;
    +------------------------------------+
    | group_concat(sname order by sname) |
    +------------------------------------+
    | Alice                              |
    | Ace                                |
    +------------------------------------+
    2 rows in set, 1 warning (0.00 sec)


TiDB master:

    mysql> set @@group_concat_max_len=5;
    Query OK, 0 rows affected (0.00 sec)
    
    mysql> select group_concat(sname order by sname) from t1 group by cid;
    +------------------------------------+
    | group_concat(sname order by sname) |
    +------------------------------------+
    | Alice                              |
    | Ace,                               |
    +------------------------------------+
    2 rows in set, 1 warning (0.01 sec)

The result of TiDB has an extra comma.

Actually, we should add "truncated" flag for each group instead of use one flag for all groups. For each group, if got truncated flag, we add the last separator. Otherwise, donot add.

**Comment by guo-shaoge** [comment]
@XuHuaiyu

**Comment by XuHuaiyu** [comment]
Have you tested the case in #23129?

**Comment by guo-shaoge** [comment]
> Have you tested the case in #23129?

Yes. The case is OK.

File: executor/aggfuncs/func_group_concat.go [hunk 1]
```diff
@@ -280,6 +280,10 @@ type topNRows struct {
 	currSize  uint64
 	limitSize uint64
 	sepSize   uint64
+	// If (separator, row) pair is poped from heap, we need to append separator in concat()
+	// In the following example, (sep2, str3) pair is poped, but partial sep2 need to be appended.
+	// eg: "str1 sep1 str2 sep2 str3" -> "str1 sep1 str2 se"
```
Comment (XuHuaiyu) [inline executor/aggfuncs/func_group_concat.go:285]: 1. sep1 --> sep
2. sep2 --> sep
3. what's the group_concat_length of this example

File: executor/aggfuncs/func_group_concat.go [hunk 2]
```diff
@@ -280,6 +280,10 @@ type topNRows struct {
 	currSize  uint64
 	limitSize uint64
 	sepSize   uint64
+	// If (separator, row) pair is poped from heap, we need to append separator in concat()
+	// In the following example, (sep2, str3) pair is poped, but partial sep2 need to be appended.
```
Comment (XuHuaiyu) [inline executor/aggfuncs/func_group_concat.go:284]: (sep2, str3) --> (sep, str3)
partial sep2 --> part of the second sep

**Review by XuHuaiyu — COMMENTED** [review]
Please refine the release note to describe what you fixed in this commit, not the phenomenon of the bug.

**Comment by guo-shaoge** [comment]
@lzmhhh123 @XuHuaiyu  PTAL

**Pull request merged**